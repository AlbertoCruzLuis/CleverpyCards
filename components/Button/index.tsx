import classNames from "classnames"
import type { HTMLProps, MouseEventHandler } from "react"
import { forwardRef } from "react"

export type ButtonVariant =
  | "primary"
  | "secondary"
  | "tertiary"

export type ButtonSize = "large" | "small"

export type ButtonProps = Omit<HTMLProps<HTMLButtonElement>, "size"> & {
  children?: React.ReactNode
  className?: string
  onClick?: MouseEventHandler<HTMLButtonElement>
  variant?: ButtonVariant
  size?: ButtonSize
  icon?: any
}

const buttonVariant = {
  primary: "bg-pink-1000 hover:bg-pink-500 text-white",
  secondary: "border border-pink-1000 hover:border-pink-400 hover:text-pink-400 text-pink-1000",
  tertiary: "text-white"
}

const buttonSize = {
  large: "px-8 py-2",
  small: "px-4 py-2"
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button (
    {
      children,
      className,
      onClick,
      variant = "primary",
      size = "small",
      icon
    }: ButtonProps,
    ref
  ) {
    const cn = classNames(className, {
      [buttonVariant.primary]: variant === "primary",
      [buttonVariant.secondary]: variant === "secondary",
      [buttonVariant.tertiary]: variant === "tertiary",
      [buttonSize.large]: size === "large",
      [buttonSize.small]: size === "small"
    })

    return (
      <button
        className={cn}
        onClick={onClick}
        type="button"
        ref={ref}
      >
        {icon && icon}
        {children}
      </button>
    )
  }
)
