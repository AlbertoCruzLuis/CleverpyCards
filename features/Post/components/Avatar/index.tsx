import { useAvatar } from "features/Post/hooks/useAvatar"
import parse from "html-react-parser"

export interface IAvatarProps {
  seed: string,
  size?: number,
}

export const Avatar = ({ seed, size = 32 }: IAvatarProps) => {
  const { avatar } = useAvatar({ seed })

  return (
    <div style={{
      width: size,
      height: size
    }}>
      { avatar && parse(avatar) }
    </div>
  )
}
