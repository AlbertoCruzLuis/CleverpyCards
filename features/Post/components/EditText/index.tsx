import classNames from "classnames"
import { useEffect, useRef, useState } from "react"

interface IEditTextProps {
  text: string,
  children?: React.ReactNode,
  className?: string,
  updateText: (text: string) => void
}

export const EditText = ({
  text,
  className = "",
  updateText,
  children
}: IEditTextProps) => {
  const [toggle, setToggle] = useState(false)
  const [textValue, setTextValue] = useState(text)
  const textAreaRef = useRef<HTMLTextAreaElement>(null)
  const styleWhenNotText = "w-full h-8 bg-red-100 rounded-sm"
  const containerStyle = classNames({
    [styleWhenNotText]: !text.length
  })

  useEffect(() => {
    if (toggle) {
      textAreaRef.current?.focus()
    }
  }, [toggle])

  if (toggle) {
    return (
      <textarea
        ref={textAreaRef}
        className={classNames(className, "resize-y rounded-md appearance-none p-1 overflow-hidden border-2 focus:border-blue-400 outline-none")}
        value={textValue}
        rows={3}
        onChange={(e) => setTextValue(e.target.value)}
        onKeyDown={(event) => {
          if (event.key === "Enter" || event.key === "Escape") {
            setToggle(false)
            updateText(textValue)
          }
        }}
        onBlur={() => {
          setToggle(false)
          updateText(textValue)
        }}
      />
    )
  } else {
    return (
      <div className={containerStyle} onClick={() => setToggle(true)}>
        {children}
      </div>
    )
  }
}
