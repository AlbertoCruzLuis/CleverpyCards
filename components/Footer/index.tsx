import { CustomLink } from "components/CustomLink"
import { Logo } from "components/Logo"
import { BsGithub, BsTwitter } from "react-icons/bs"

export const Footer = () => {
  const year = new Date().getFullYear()
  return (
    <footer className="p-8 bg-[#1D0327]">
      <div className="mx-auto xl:container">
        <div className="grid grid-cols-4 gap-8 xs:grid-cols-1 sm:grid-cols-2 md:grid-cols-2">
          <div className="flex flex-col gap-4">
            <CustomLink href="/">
              <Logo width="90" height="90" />
            </CustomLink>
            <p className="max-w-sm font-semibold text-white">Cleverpy Cards</p>
            <div className="flex gap-4">
              <CustomLink href="" className="p-2 border-2 border-neutral-700">
                <BsTwitter size="1.2rem" className="transition-all text-neutral-400 hover:text-white" />
              </CustomLink>
              <CustomLink href="" className="p-2 border-2 border-neutral-700">
                <BsGithub size="1.2rem" className="transition-all text-neutral-400 hover:text-white" />
              </CustomLink>
            </div>
          </div>
          <div>
            <h2 className="mb-6 text-sm font-semibold text-white uppercase">Resources</h2>
            <ul className="text-neutral-400">
              <li className="mb-4">
                <CustomLink href="/" className="hover:underline">Home</CustomLink>
              </li>
            </ul>
          </div>
          <div>
            <h2 className="mb-6 text-sm font-semibold text-white uppercase">Follow us</h2>
            <ul className="text-neutral-400">
              <li className="mb-4">
                <CustomLink href="" className="hover:underline ">Github</CustomLink>
              </li>
              <li>
                <CustomLink href="" className="hover:underline">Twitter</CustomLink>
              </li>
            </ul>
          </div>
          <div>
            <h2 className="mb-6 text-sm font-semibold text-white uppercase">Legal</h2>
            <ul className="text-neutral-400">
              <li className="mb-4">
                <CustomLink href="/" className="hover:underline">Privacy Policy</CustomLink>
              </li>
              <li>
                <CustomLink href="#" className="hover:underline">Terms &amp; Conditions</CustomLink>
              </li>
            </ul>
          </div>
        </div>
        <hr className="my-6 border-neutral-800/50 sm:mx-auto lg:my-8" />
        <div className="sm:flex sm:items-center sm:justify-between">
          <span className="text-sm text-neutral-400 sm:text-center">© {year} <CustomLink href="/" className="hover:underline">Cleverpy™</CustomLink>. All Rights Reserved.
          </span>
        </div>
      </div>
    </footer>
  )
}
