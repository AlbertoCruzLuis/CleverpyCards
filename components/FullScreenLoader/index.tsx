import { Background } from "components/Background"
import { useAuth } from "hooks/useAuth"
import { useRouter } from "next/router"
import { useEffect, useState } from "react"
import ScaleLoader from "react-spinners/ScaleLoader"

interface IFullScreenLoaderProps {
  children: React.ReactNode
}

export const FullScreenLoader = ({ children }: IFullScreenLoaderProps) => {
  const { isAuth } = useAuth()
  const [isLoading, setLoading] = useState(true)
  const router = useRouter()

  const showLoader = () => {
    const isLoginPage = router.pathname === "/login"
    const isShowLoader = isAuth === isLoginPage
    setLoading(isShowLoader)
  }

  useEffect(() => {
    setTimeout(() => {
      showLoader()
    }, 800)
  }, [router, isAuth])

  if (!isLoading) {
    return (
      <>
        {children}
      </>
    )
  }

  return (
    <Background>
      <div className="flex items-center justify-center h-screen">
        <ScaleLoader color="#fff" />
      </div>
    </Background>
  )
}
