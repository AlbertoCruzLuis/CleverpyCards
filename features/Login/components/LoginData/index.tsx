import { Button } from "components/Button"
import { ErrorMessage, Field, useFormikContext } from "formik"
import toast from "react-hot-toast"

export const styles = {
  firstLabel: "text-gray-200 text-sm font-semibold mb-2 block",
  label: "text-neutral-200 text-sm font-semibold mb-2 block",
  field: "bg-neutral-100 text-gray-700 focus:outline-none focus:shadow-outline border border-gray-300 rounded py-2 px-4 block w-full appearance-none",
  socialField: "bg-neutral-100 text-gray-700 focus:outline-none focus:shadow-outline border border-gray-300 rounded-r-md py-2 px-4 block w-full appearance-none",
  errorMessage: "text-red-500 text-sm"
}

export const LoginData = () => {
  const { handleSubmit, errors } = useFormikContext()

  const submitForm = async () => {
    try {
      const isError = Object.keys(errors).length !== 0

      if (isError) {
        toast.error("Please fill all required fields")
      }

      handleSubmit()
    } catch (error) {
      console.error(error)
    }
  }

  return (
    <div className="flex flex-col items-center gap-10 bg-[#1D0327] p-8 rounded-md">
      <div className="flex flex-col gap-4">
        <div>
          <label className={styles.label} htmlFor="email">Email</label>
          <Field id="email" className={styles.field} type="text" name="email" placeholder="alberto@cleverpy.com" />
          <ErrorMessage className={styles.errorMessage} name="email" component="div" />
        </div>
        <div>
          <label className={styles.label} htmlFor="password">Password</label>
          <Field id="password" className={styles.field} type="password" name="password" placeholder="alberto" />
          <ErrorMessage className={styles.errorMessage} name="password" component="div" />
        </div>
      </div>
      <Button
        className="w-full rounded-sm"
        variant="primary"
        onClick={submitForm}
      >
        <span>Sign in</span>
      </Button>
    </div>
  )
}
