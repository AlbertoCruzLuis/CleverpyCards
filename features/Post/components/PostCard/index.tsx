import { Avatar } from "features/Post/components/Avatar"
import { usePost } from "features/Post/hooks/usePost"
import { motion } from "framer-motion"
import { BsTrash } from "react-icons/bs"
import { EditText } from "../EditText"

interface IPostCardProps {
  id: number,
  title: string,
  description: string,
  user: {
    id: number
  }
}

export const PostCard = ({
  id,
  title,
  description,
  user
}: IPostCardProps) => {
  const { updatePost, removePost } = usePost()

  return (
    <motion.div
      className="relative flex flex-col justify-between gap-4 p-4 rounded-sm bg-neutral-200"
      initial={{ opacity: id * 0.1 }}
      animate={{ opacity: 1 }}
      exit={{ opacity: 0 }}
    >
      <span className="absolute font-bold top-2 right-2">#{id}</span>
      <div className="flex flex-col gap-2">
        <div className="flex items-center gap-2">
          <Avatar seed={`${user.id}`} />
          <span>User {user.id}</span>
        </div>
        <div className="flex flex-col gap-2">
          <EditText
            className="text-xl font-bold"
            text={title}
            updateText={(textUpdated: string) => updatePost({
              id,
              userId: user.id,
              title: textUpdated,
              body: description
            })}
          >
            <h3 className="text-xl font-bold">{title}</h3>
          </EditText>
          <EditText
            className="text-sm"
            text={description}
            updateText={(textUpdated: string) => updatePost({
              id,
              userId: user.id,
              title,
              body: textUpdated
            })}
          >
            <p className="text-sm">{description}</p>
          </EditText>
        </div>
      </div>
      <div className="flex gap-4">
        <button onClick={() => removePost(id)}>
          <BsTrash className="text-gray-600 transition-all hover:rotate-12" />
        </button>
      </div>
    </motion.div>
  )
}
