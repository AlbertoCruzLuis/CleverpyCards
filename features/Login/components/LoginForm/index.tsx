import { Form, Formik } from "formik"
import { useAuth } from "hooks/useAuth"
import { loginSchema } from "utils/validation/loginSchema"
import { LoginData } from "../LoginData"

export interface IFormValues {
  email: string,
  password: string
}

export const LoginForm = () => {
  const { logIn } = useAuth()
  const initialFormValues = {
    email: "",
    password: ""
  }

  const handleSubmit = async (values: IFormValues) => {
    const formatedData = {
      email: values.email,
      password: values.password
    }

    logIn(formatedData)
  }
  return (
    <div>
      <Formik
        initialValues={initialFormValues}
        validationSchema={loginSchema}
        onSubmit={(values) => handleSubmit(values)}
      >
        {() => {
          return (
            <Form>
              <LoginData />
            </Form>
          )
        }}
      </Formik>
    </div>
  )
}
