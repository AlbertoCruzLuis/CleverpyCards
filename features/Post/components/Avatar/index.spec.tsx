import { Avatar } from "features/Post/components/Avatar"
import { customRender } from "test"

describe("Avatar", () => {
  it("renders correctly", () => {
    const { container } = customRender(<Avatar seed="123" />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
