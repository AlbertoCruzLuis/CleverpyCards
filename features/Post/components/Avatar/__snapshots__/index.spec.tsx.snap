// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders correctly 1`] = `
<div
  style="width: 32px; height: 32px;"
/>
`;
